batch-processing/pipeline.go

```diff
@@ -0,0 +1,353 @@
+	defer wg.Done()
+	for {
+		row, ok := <-p.channels[CONVERT]
+		if !ok {
+			break
+		}
+
+		if row.Err != nil {
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		file, err := os.Open(row.Output)
+
+		if err != nil {
+			row.Err = err
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		// Upload to S3
+		path := strings.Split(row.Output, "/")
+		key := path[len(path)-1]
+
+		err = UploadToS3WithBackoff(file, key, p.config.Aws, p.maxRetries)
+
+		if err != nil {
+			row.Err = err
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		row.S3url = fmt.Sprintf("https://%s.s3.amazonaws.com/%s", p.config.Aws.s3bucket, key)
+
+		p.channels[UPLOAD] <- row
+	}
+}
+
+func (p *Pipeline) Write(done chan bool) {
+
+	f, err := os.Create(p.config.OutputFilepath)
+	if err != nil {
+		log.Fatalf("error creating output file: %v", err)
+	}
+	defer f.Close()
+
+	w := csv.NewWriter(f)
+	defer w.Flush()
+	w.Write(OutputHeader)
+	var failedW *csv.Writer
+
+	// Create csv for failed output if param is set
+	if p.config.FailedOutputFilepath != "" {
+		failedF, err := os.Create(p.config.FailedOutputFilepath)
+		if err != nil {
+			log.Fatalf("error creating failed output file: %v", err)
+		}
+		defer failedF.Close()
+
+		failedW = csv.NewWriter(failedF)
+		defer failedW.Flush()
+		failedW.Write(FailedOutputHeader)
+	}
+
+	for {
+		row, ok := <-p.channels[UPLOAD]
+
+		if !ok {
+			break
+		}
+		rowErr := ""
+		// only add to failed output if row has an error, but not a duplicate error or image exists error
+		if row.Err != nil {
+			if p.config.FailedOutputFilepath != "" && row.Err != ErrDuplicateURL && row.Err != ErrImageExistsInS3 {
+				failedW.Write([]string{row.Url})
+			}
+			rowErr = row.Err.Error()
+		}
+
+		w.Write([]string{row.Url, row.Input, row.Output, row.S3url, rowErr})
+	}
+	done <- true
+}
+
+func (p *Pipeline) taskPicker(task string) Task {
+	switch task {
+	case DOWNLOAD:
+		return p.Download
+	case CONVERT:
+		return p.Convert
+	case UPLOAD:
+		return p.Upload
+	}
+
+	return nil
+}
+
+func (p *Pipeline) Execute() error {
+
+	start := time.Now()
+
+	done := make(chan bool)
+	go p.Write(done)
+
+	// Start the workers
+	wgTasks := &sync.WaitGroup{}
+	for _, task := range TASKS {
+		wgTasks.Add(1)
+		go func(task string) {
+			var wg sync.WaitGroup
+			for i := 0; i < p.workers[task]; i++ {
+				wg.Add(1)
+				go p.taskPicker(task)(&wg)
+			}
+			wg.Wait()
+			close(p.channels[task])
+			wgTasks.Done()
+		}(task)
+	}
+
+	// Start the reader
+	csvReader, err := OpenCSVFile(p.config.InputFilepath)
+
+	if err != nil {
+		return err
+	}
+	go p.Read(csvReader)
+
+	// Wait for all tasks to finish
+	wgTasks.Wait()
+	elapsed := time.Since(start)
+
+	log.Printf("Finished in %s", elapsed)
+
+	<-done
+
+	return nil
+}
+
+func NewPipeline(config *Config) *Pipeline {
+	return &Pipeline{
+		config:     config,
+		maxRetries: 3,
+		urlMap:     &sync.Map{},
+		channels: map[string]chan *Out{
+			READ:     make(chan *Out),
+			DOWNLOAD: make(chan *Out),
```

Comment:
It's not super clear to me that every pipeline stage using the same type, `*Out`, makes for the easiest to follow and maintain code... Can you contrast the pros/cons of per-channel types vs using the same one for each?

Reply:
The reason I went with 1 type:
1. Less boilerplate code, no need to create and manage multiple types
2. Easier to setup workers as they have the same interface

Cons:
The struct keeps getting bigger as pipeline grows and more complex, e.g. for simple read tasks we would be storing a bigger than needed struct adding unnecessary complexity and memory usage.

Reply:
Sounds right - the other big con is that it makes it hard to know you're writing to the right channel. Having different types on each channel means you can't accidentally write to the convert or upload channel from the read goroutine - the typed channels ensure that each goroutine can only write its output where its output type is actually expected.